### PR TITLE
[MPS] Apply inter-layer dropout in LSTM backward and fix dropout=1 NaN

### LINKED ISSUE #190056: [MPS] Multilayer LSTM backward ignores dropout, silently corrupting training gradients
### 🐛 Describe the bug

`nn.LSTM` with `num_layers >= 2` and `dropout > 0` computes wrong gradients on
MPS. The forward saves each layer's output *before* applying inter-layer
dropout, and the backward consumes those unmasked tensors and never reapplies
the mask — so both the inter-layer gradient chain and the upper layers' weight
gradients are computed as if no dropout had happened.

Deterministic check via fixed-RNG central differences (resetting the device
RNG before every forward makes the dropout mask identical across evaluations):

```python
import torch

torch.manual_seed(0)
lstm = torch.nn.LSTM(4, 4, num_layers=2, dropout=0.5).to("mps")
lstm.train()

def loss(x):
    torch.mps.manual_seed(42)   # identical dropout mask on every call
    out, _ = lstm(x)
    return out.sum()

x = torch.randn(3, 2, 4, device="mps", requires_grad=True)
loss(x).backward()
analytic = x.grad.flatten()[0].item()
eps = 1e-2
with torch.no_grad():
    xp = x.detach().clone(); xp.flatten()[0] += eps
    xm = x.detach().clone(); xm.flatten()[0] -= eps
    numeric = ((loss(xp) - loss(xm)) / (2 * eps)).item()
print(f"analytic={analytic:.6f} numeric={numeric:.6f}")
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable and nightly 2.14.0.dev20260710:

```text
analytic=0.000132 numeric=0.006306
```

That is a 98% relative error. Running the same script with `dropout=0.0` as a
control gives `analytic=-0.007924 numeric=-0.007927` (0.04% error), so the
finite-difference setup itself is sound. Weight gradients are similarly wrong
(a two-layer `dropout=0.5` LSTM had layer-0 weight gradient analytic -9.69e-4
vs numeric +1.97e-4).

The forward stores pre-dropout layer outputs in `layersOutputs`
(https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/RnnOps.mm#L219-L256),
and the backward slices them as the lower layers' inputs and calls
`LSTMGradientsWithSourceTensor` with `mask:nil`, with no mask multiply in the
inter-layer gradient chain
(https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/RnnOps.mm#L580-L615).

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main), 2.13.0, and nightly 2.14.0.dev20260710; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

### LINKED ISSUE #190057: [MPS] Multilayer LSTM with dropout=1.0 returns all NaN
### 🐛 Describe the bug

A training-mode `nn.LSTM` with `num_layers >= 2` and `dropout=1.0` returns NaN
for every output element on MPS. The hand-rolled inter-layer dropout builds an
all-zero keep mask at p=1 and then multiplies it by `1/(1-p) = inf`, so every
masked element becomes `0 * inf = NaN`. PyTorch's defined `dropout=1` behavior
is a zero inter-layer input, not NaN — CPU returns finite output.

```python
import torch

torch.manual_seed(0)
lstm = torch.nn.LSTM(4, 4, num_layers=2, dropout=1.0)
x = torch.randn(3, 2, 4)
print(lstm(x)[0].isfinite().all())          # CPU: True
print(lstm.to("mps")(x.to("mps"))[0].isnan().all())  # MPS: True (all NaN)
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable and nightly 2.14.0.dev20260710.

The scale constant is built unconditionally as `1.0 / (1.0 - dropout_p)`:
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/RnnOps.mm#L249-L252

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main), 2.13.0, and nightly 2.14.0.dev20260710; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @mikaylagawarecki @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
Forward saved each layer's output *before* applying inter-layer dropout, and backward consumed those unmasked tensors with `mask:nil`, so gradients were computed as if no dropout had happened. Against fixed-RNG central differences (reseeding the device RNG per forward pins the mask), a two-layer `dropout=0.5` LSTM showed 98% relative error on the input gradient, while the `dropout=0` control matched to 0.0%.

Forward now saves the post-dropout tensor (what the next layer actually consumed) and stacks the scaled dropout masks after the layer outputs inside the existing `layersOutputs` tensor, avoiding a schema change; the private forward/backward pair agree on the layout. Backward slices the masks back out and multiplies the inter-layer gradient. The input slices switch from end-relative to front-relative indexing (the tensor's first dimension doubles when dropout is active), and the backward graph key gains dropout/train.

Separately, the dropout scale was built unconditionally as `1/(1-p)`, so `dropout=1.0` produced `0 * inf = NaN` for every output element; the scale is now 0 at p=1, matching the defined zero-inter-layer-input behavior (new test compares against CPU, which is deterministic at p=1).

After the fix the finite-difference check matches to 0.08% relative error. The pre-existing RNN suite passes.

Fixes #190056
Fixes #190057

*This PR was authored with assistance from Claude.*

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 1]
```diff
@@ -611,6 +620,23 @@
...
+          // The gradient just computed is w.r.t. layer i's post-dropout
+          // input; multiply by the scaled mask saved by forward to get the
+          // gradient w.r.t. layer i-1's raw output.
+          MPSGraphTensor* maskTensor = [mpsGraph sliceTensor:layersOutputsTensor
+                                                   dimension:0
+                                                       start:(num_layers - 1) + (i - 1)
+                                                      length:1
+                                                        name:nil];
+          if (is_macos_14_4_or_newer) {
+            maskTensor = [mpsGraph identityWithTensor:maskTensor name:nil];
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:633]: Mental note to me to check if/when this has been fixed so we can get rid of it appropriately.

File: test/test_mps.py [hunk 2]
```diff
@@ -15036,6 +15036,57 @@ def test_lstm_backward(self, device="mps", dtype=torch.float32):
             for test_options in self.LSTM_TEST_CASES:
                 self._lstm_helper(num_layers=num_layers, dtype=dtype, device=device, backward=True, **test_options)
 
+    # Regression test for https://github.com/pytorch/pytorch/issues/190057:
```
Comment (jhavukainen) [inline test/test_mps.py:15039]: Checked that dropout doesn't seem to get exercised in the OpInfo or Module tests shared between backends. Presumably because there has been no explicit contract for the RNGs to agree between implementations so asking for the dropout to agree is meaningless.

So these tests should exist and are in the right place.

**Review by jhavukainen — APPROVED** [review]
Looks clean and correct to me. It addresses the issues and everything seems to be gated correctly and in line with what the other backends would do. 

Tests seem to be in the right place at least for now, no pre-existing tests covering this case in the shared tests as far as I can tell. And probably due to the RNGs differing between backends we don't want go through implementing those in this PR at least.

**Comment by jhavukainen** [comment]
Tests have cleared.

@pytorchbot merge
